src/common/common_service/src/observer_manager.rs

```diff
@@ -88,18 +90,66 @@ where
         }
     }
 
+    async fn wait_init_notification(&mut self) {
+        let mut notification_vec = Vec::new();
+        let mut init_notification = loop {
+            // notification before init notification must be received successfully.
+            let notification = self.rx.message().await.unwrap().unwrap();
+            if !matches!(notification.info.as_ref().unwrap(), &Info::Snapshot(_)) {
+                notification_vec.push(notification);
+            } else {
+                break notification;
+            }
+        };
+
+        let Info::Snapshot(info) = init_notification.info.as_ref().unwrap() else {
+            unreachable!();
+        };
+
+        let SnapshotVersion {
+            catalog_version,
+            parallel_unit_mapping_version,
+            worker_node_version,
+        } = info.version.clone().unwrap();
+
+        notification_vec.retain_mut(|notification| match notification.info.as_ref().unwrap() {
+            Info::Database(_)
+            | Info::Schema(_)
+            | Info::Table(_)
+            | Info::Source(_)
+            | Info::Sink(_)
+            | Info::Index(_)
+            | Info::View(_)
+            | Info::User(_) => notification.version > catalog_version,
+            Info::ParallelUnitMapping(_) => notification.version > parallel_unit_mapping_version,
+            Info::Node(_) => notification.version > worker_node_version,
+            Info::HummockVersionDeltas(version_delta) => {
+                version_delta.version_deltas[0].id > info.hummock_version.as_ref().unwrap().id
+            }
+            Info::HummockSnapshot(_) => true,
+            Info::Snapshot(_) => unreachable!(),
+        });
+
+        let init_notification_version = init_notification.version - notification_vec.len() as u64;
```

Comment:
It's quite weird to modify version of init and buffered notifications here. Why don't just use `SnapshotVersion` in init notification separately? Then we can handle buffered notification without changing their versions.

Reply:
If the buffered notification need to be handled, it will have a larger version than the one in snapshot. ie:
received: <catalog_10, user_11, user_12, catalog_13, snapshot_11(catalog_11, user_11)_14,...>
filtered: notifications => [user_12, catalog_13],  snapshot => [catalog_11, user_11]
handle_init: catalog: 11,  user: 11
handle_notification: catalog: 13, user: 12

Here I mean you should init catalog with version 11 in snapshot and user with 11, NOT using the version of the notification(which is 14). It's naturally that you will have different versions for different managers in this PR. And the buffered notification won't break the order if you filtered the notification correctly.

Reply:
Agree. I will edit it later.